game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java

```diff
@@ -92,84 +89,77 @@
-                validTargets.removeAll(casualtiesSoFar);
-                if (!validTargets.isEmpty()) {
-                  dice =
-                      DiceRoll.rollAa(
-                          validTargets,
-                          firingGroup,
-                          allEnemyUnitsAliveOrWaitingToDie,
-                          allFriendlyUnitsAliveOrWaitingToDie,
-                          bridge,
-                          battleSite,
-                          defending);
-                  if (!headless) {
-                    SoundUtils.playFireBattleAa(firingPlayer, aaType, dice.getHits() > 0, bridge);
-                  }
+    for (final FiringGroup groupAndTarget : groupsAndTargets) {
+      final Collection<Unit> validTargets = groupAndTarget.getValidTargets();
+      final Collection<Unit> firingGroup = groupAndTarget.getFiringGroup();
+      final String aaType = groupAndTarget.getType();
+
+      final IExecutable rollDice =
+          new IExecutable() {
+            private static final long serialVersionUID = 6435935558879109347L;
+
+            @Override
+            public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+              validTargets.removeAll(casualtiesSoFar);
+              if (!validTargets.isEmpty()) {
+                dice =
+                    DiceRoll.rollAa(
+                        validTargets,
+                        firingGroup,
+                        allEnemyUnitsAliveOrWaitingToDie,
+                        allFriendlyUnitsAliveOrWaitingToDie,
+                        bridge,
+                        battleSite,
+                        defending);
+                if (!headless) {
+                  SoundUtils.playFireBattleAa(firingPlayer, aaType, dice.getHits() > 0, bridge);
                 }
               }
-            };
-        final IExecutable selectCasualties =
-            new IExecutable() {
-              private static final long serialVersionUID = 7943295620796835166L;
+            }
+          };
+      final IExecutable selectCasualties =
+          new IExecutable() {
+            private static final long serialVersionUID = 7943295620796835166L;
 
-              @Override
-              public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
-                if (!validTargets.isEmpty()) {
-                  final CasualtyDetails details =
-                      selectCasualties(validTargets, firingGroup, bridge, aaType);
-                  battle.markDamaged(details.getDamaged(), bridge);
-                  casualties = details;
-                  casualtiesSoFar.addAll(details.getKilled());
-                }
+            @Override
+            public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+              if (!validTargets.isEmpty()) {
+                final CasualtyDetails details =
+                    selectCasualties(validTargets, firingGroup, bridge, aaType);
```

Comment:
This could also take FiringGroup directly. Then it just need two params (FiringGroup and bridge).